proposals/sealed-inline-classes.md

```diff
@@ -0,0 +1,587 @@
+
+## Open function on top, override in bottom
+
+```kotlin
+@JvmInline
+sealed value class I1 {
+    open fun str(): String = "I1"
+}
+
+@JvmInline
+sealed value class I2 : I1()
+
+sealed object O1: I1() {
+    override fun str(): String = "O1"
+}
+
+@JvmInline
+value class I3(val a: Any?): I2() {
+    override fun str(): String = "I3"
+}
+
+value object O2: I2()
+```
+
+The compiler generates the following:
+
+```kotlin
+fun `str-impl`(value: Any?): String {
+    when (value) {
+        is O1 -> return (value as O1).str()
+        is O2 -> return "I1"
+        else -> return I3.str-impl(value)
+    }
+}
+
+// In I2
+// Nothing
+
+// In I3
+fun `str-impl`(value: Any?): String = "I3"
+```
+
+Note, that we simply copy original body if there is no override in children.
+
+## Open function on top, override in middle and bottom
+
+```kotlin
+@JvmInline
+sealed value class I1 {
+    open fun str(): String = "I1"
+}
+
+@JvmInline
+sealed value class I2 : I1() {
+    override fun str(): String = "I2"
+}
+
+value object O1: I1()
+
+@JvmInline
+value class I3(val a: Any?): I2() {
+    override fun str(): String = "I3"
+}
+
+value object O2: I2()
+```
+
+Generated code:
+
+```kotlin
+// In I1
+fun `str-impl`(value: Any?): String {
+    when (value) {
+        is O1 -> return "I1"
+        else -> return I2.str-impl(value)
+    }
+}
+
+// In I2
+fun `str-impl`(value: Any?): String {
+    when (value) {
+        is O2 -> return "I2"
+        else -> return I3.str-impl(value)
+    }
+}
+
+// In I3
+fun `str-impl`(value: Any?): String = "I3"
+```
+
+## Open function on top, override in middle
+
+```kotlin
+@JvmInline
+sealed value class I1 {
+    open fun str(): String = "I1"
+}
+
+@JvmInline
+sealed value class I2 : I1() {
+    override fun str(): String = "I2"
+}
+
+value object O1: I1()
+
+@JvmInline
+value class I3(val a: Any?): I2()
+
+value object O2: I2()
+```
+
+becomes:
+
+```kotlin
+/// In I1
+fun `str-impl`(value: Any?): String {
+    when (value) {
+        is O1 -> return "I1"
+        else -> return I2.str-impl(value)
+    }
+}
+
+// In I2
+fun `str-impl`(value: Any?): String = "I2"
+```
+
+## Open function on top, no override
+
+```kotlin
+@JvmInline
+sealed value class I1 {
+    open fun str(): String = "I1"
+}
+
+@JvmInline
+sealed value class I2 : I1()
+
+value object O1: I1()
+
+@JvmInline
+value class I3(val a: Any?): I2()
+
+value object O2: I2()
+```
+
+The simplest case:
+
+```kotlin
+fun `str-impl`(value: Any?): String = "I1"
+```
+
+## Open function in middle, no override
+
+```kotlin
+@JvmInline
+sealed value class I1
+
+@JvmInline
+sealed value class I2 : I1() {
+    open fun str(): String = "I2"
+}
+
+value object O1: I1()
+
+@JvmInline
+value class I3(val a: Any?): I2()
+
+value object O2: I2()
+```
+
+Here, we need to generate synthetic method in `I1`, since we only box to `I1`
+and we need to get to `I2`.
+
+```kotlin
+// In I1
+synthetic fun `str-impl`(value: Any?): String = I2.`str-impl`(value)
```

Comment:
`synthetic` has been used inconsistently in the examples.
See the following lines or the definition of the `$value` property.